File: src/manchester_weekly/scripts/week1/static_tf.py
#Node file, must go in src or scripts if usied whithin a cmake env
import rclpy
from rclpy.node import Node
from tf2_ros import StaticTransfromBroadcster
from geometry_msgs.msg import TransformStamped
import transforms3d

class FramePublisher(Node):
    def __init__(self):
        super().__init('frame publisher')
        self.static_br1 = StaticTransfromBroadcster(self)
        self.static_br2 = StaticTransfromBroadcster(self)

        t = TransformStamped()
        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = 'world'
        t.child_frame_id = 'robot_3'
        t.transform.translation.x = -2.0
        t.transform.translation.y = -1.0
        t.transform.translation.z = 0.0
        q = transforms3d.euler.euler2quat(0,0,0) # input euler2quat(roll, pitch, yaw), output quaternion (w,x,y,z)
        t.transform.rotation.w = q[0]
        t.transform.rotation.x = q[1]
        t.transform.rotation.y = q[2]
        t.transform.rotation.z = q[3]
        
        t2 = TransformStamped()
        t2.header.stamp = self.get_clock().now().to_msg()
        t2.header.frame_id = 'robot_2'
        t2.child_frame_id = 'robot_4'
        t2.transform.translation.x = 1.0
        t2.transform.translation.y = 1.0
        t2.transform.translation.z = 1.0
        q2 = transforms3d.euler.euler2quat(1.57,1.57,0) # input euler2quat(roll, pitch, yaw), output quaternion (w,x,y,z)
        t2.transform.rotation.w = q[0]
        t2.transform.rotation.x = q[1]
        t2.transform.rotation.y = q[2]
        t2.transform.rotation.z = q[3]

        self.static_br1.sendTransform(t)
        self.static_br2.sendTransform(t2)
        
        
def main(args=None):
    rclpy.init(args=args)   
    node = FramePublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        if rclpy-ok():
            rclpy.shutdown()
            node.destroy_node()

if __name__='__main__':
    main()

File: src/manchester_weekly/src/week1/static_tf.cpp
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2_ros/static_transform_broadcaster.h"
#include "tf2/LinearMath/Quaternion.h"

class StaticFramePublisher : public rclcpp::Node {
public:
    StaticFramePublisher() : Node("static_frame_publisher") {
        static_br_ = std::make_unique<tf2_ros::StaticTransformBroadcaster>(*this);

        // --- Static Transform 1 ---
        geometry_msgs::msg::TransformStamped t;
        t.header.stamp = this->get_clock()->now();
        t.header.frame_id = "world";
        t.child_frame_id = "robot_3";
        t.transform.translation.x = -2.0;
        t.transform.translation.y = -1.0;
        t.transform.translation.z = 0.0;

        tf2::Quaternion q;
        q.setRPY(0, 0, 0);
        t.transform.rotation.x = q.x();
        t.transform.rotation.y = q.y();
        t.transform.rotation.z = q.z();
        t.transform.rotation.w = q.w();

        // --- Static Transform 2 ---
        geometry_msgs::msg::TransformStamped t2;
        t2.header.stamp = this->get_clock()->now();
        t2.header.frame_id = "robot_2"; // Matches your Python logic
        t2.child_frame_id = "robot_4";
        t2.transform.translation.x = 1.0;
        t2.transform.translation.y = 1.0;
        t2.transform.translation.z = 1.0;

        tf2::Quaternion q2;
        q2.setRPY(1.57, 1.57, 0); 
        t2.transform.rotation.x = q2.x();
        t2.transform.rotation.y = q2.y();
        t2.transform.rotation.z = q2.z();
        t2.transform.rotation.w = q2.w();

        // Send all static transforms as a vector
        static_br_->sendTransform({t, t2});
    }

private:
    std::unique_ptr<tf2_ros::StaticTransformBroadcaster> static_br_;
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<StaticFramePublisher>());
    rclcpp::shutdown();
    return 0;
}
